python/cudf/cudf/core/groupby/groupby.py

```diff
@@ -701,6 +702,45 @@ def cummax(self):
         """Get the column-wise cumulative maximum value in each group."""
         return self.agg("cummax")
 
+    def shift(self, periods=1, freq=None, axis=0, fill_value=None):
+        """
+        Shift each group by `periods` positions.
+
+        Parameter "freq" is unsupported.
+
+        Parameters
+        ----------
+        periods : int, default 1
+        Number of periods to shift.
+
+        freq : str, unsupported
+
+        axis : 0, axis to shift
+        Shift direction. Only row-wise shift is supported
+
+        fill_value : optional
+        The scalar value to use for newly introduced missing values. Should
+        match the dtype of columns to fill.
```

Comment:
(how) Does Pandas handle the case where the columns aren't all of the same dtype?

Reply:
Coerced into object dtype:
```
>>> df = pd.DataFrame({'a': [1, 1, 2], 'b': [4, 5, 6]})
>>> df['b'].dtype
dtype('int64')
>>> res = df.groupby('a').shift(periods=1, fill_value='x')
>>> res['b'].dtype
dtype('O')
```

Reply:
Pandas does not really enforce that `fill_value` has to be the same dtype as the columns to fill, in case they are different, the column will be coerced into object dtype:
```
>>> df = pd.DataFrame({'a': [1, 1, 2], 'b': [4, 5, 6], 'c': ['x', 'y', 'z']})
>>> res = df.groupby('a').shift(periods=1, fill_value='m')
>>> res['b'].dtype
dtype('O')
```

Reply:
Pandas does not seem to accept multiple `fill_values` (one for each column) either.
```
>>> df = pd.DataFrame({'a': [1,1,2], 'b': [3, 4, 5], 'c': ['x', 'y', 'z']})
>>> df
   a  b  c
0  1  3  x
1  1  4  y
2  2  5  z
>>> df.groupby('a').shift(1, fill_value=[1, 'm'])
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
  File "/raid/wangm/dev/rapids/compose/etc/conda/cuda_11.2/envs/rapids/lib/python3.7/site-packages/pandas/core/groupby/groupby.py", line 2722, in shift
    if freq is not None or axis != 0 or not isna(fill_value):
ValueError: The truth value of an array with more than one element is ambiguous. Use a.any() or a.all()**
```

Reply:
Hmm, then I think our behaviour is reasonable. Although ideally it would "just work" if one supplied `fill_value=0` and the dtypes were all numeric, but possibly different. That could be a TODO for the future though.

Reply:
You mean:
```
>>> df = cudf.DataFrame({'a':[1,1,2], 'b':[2, 3, 4], 'c':[3.14, 1.71, 0.0]})
>>> df
   a  b     c
0  1  2  3.14
1  1  3  1.71
2  2  4  0.00
>>> df.groupby('a').shift(1, fill_value=0)
   b     c
a         
1  0  0.00
1  2  3.14
2  0  0.00
```
This currently just works thanks to:
```python
        for val, col in zip(fill_values, values._columns):
            d_slrs.append(as_device_scalar(val, dtype=col.dtype))
```

Reply:
Sweet!

Reply:
Added a test case for mixed numeric types.